### ISSUE TITLE
improve performance

### ISSUE THREAD

**Issue by matryer** [matryer/way#1]
@dahernan talked about a way to make this way quicker (using radix trees) -- Let's do it.

**Comment by matryer** [comment]
@ole108 How are you getting on with this?

**Comment by ole108** [comment]
@matryer  Sorry Mat but I got a bit stuck with it.
I tried first to take [httptreemux](https://github.com/dimfeld/httptreemux) (it already has got many of the nice things we want and the original from Julien Schmidt doesn't got) and cut code from there and paste it into way. Unfortunately it was quite impossible to make ends meet.

Then I forked httptreemux instead and tried to refactor it until the interface would match. But I got stuck there, too.

A month ago I started a new job and consequently don't have got much time for it now.
I suppose one would have to only look at code from other routers but shouldn't try to copy much.
I really hoped it would be much easier.

**Comment by matryer** [comment]
No problem Ole - it's just an interesting side problem. No real pressure to get it done. 

Thanks anyway
Mat

**Comment by awulkan** [comment]
Sorry for resurrecting an old issue, but does anyone have a benchmark of the current state of the router? How bad/good is the performance right now?

**Comment by matryer** [comment]
Thanks for your comment.

Router performance isn’t considered an important metric in the wider
context of where it is used in real world applications, so we don’t give it
much time or attention. There is an idea to use a tree to lookup the routes
which would be faster but it’s more of an intellectual exercise rather than
an important priority.

Mat

**Comment by awulkan** [comment]
Yes, I totally understand that routing performance isn't the most time consuming part in applications. I was mainly asking since you brought it up in this issue. I'm new to Go, but it would be an interesting project for me to maybe work on. :)

Also, thanks for the quick response.

**Comment by matryer** [comment]
Well if you fancy having a look, the idea is to build a version of this that uses [a radix tree](https://en.wikipedia.org/wiki/Radix_tree) for lookups. And benchmark it to see what difference it makes. Purely academic exercise though :)

**Comment by awulkan** [comment]
@matryer Ye, I'm already kinda done. :)
I started with this project as a base, but pretty much ended up rewriting everything. Still working on it, but so far the performance is pretty good. The only thing that appears to lower the performance is when assigning parameters to the context.

Unfortunately I don't think it would be a good idea to merge my code with this project since it's so different. I mean, it looks just like your router from a user's perspective, but the router's code is almost fully replaced with my own.

I might post it on github someday if anyone is interested.

**Comment by ole-optiopay** [comment]
@matryer @awulkan IMHO this is the only way.
As long as the API stays the same and the new code isn't too complex it should be fine.
I once tried to solve this issue with smaller changes but it didn't add up...

**Comment by ofabricio** [comment]
[this](https://github.com/gbrlsnchs/radix#building-a-dynamic-tree) might be useful

**Comment by BackAged** [comment]
is someone still working on `improve performance`?

**Comment by davidmdm** [comment]
pinging for visibility. Is there still interest in adding a radix tree implementation to `way`?

**Comment by davidmdm** [comment]
@matryer ? Any interest?
